pkgs/os-specific/darwin/cctools/llvm.nix

```diff
@@ -0,0 +1,89 @@
+# Create a cctools-compatible bintools that uses equivalent tools from LLVM in place of the ones
+# from cctools when possible.
+
+{ lib, stdenv, makeWrapper, cctools-port, llvmPackages, useLLD ? false }:
+
+let
+  cctoolsVersion = lib.getVersion cctools-port;
+  llvmVersion = llvmPackages.release_version;
+
+  # A compatible implementation of `otool` was not added until LLVM 13.
+  useLLVMOtool = lib.versionAtLeast llvmVersion "13";
+
+  # Older versions of `strip` cause problems for the version of `codesign_allocate` available in
+  # the version of cctools in nixpkgs. The version of `codesign_allocate` in cctools-1005.2 does
+  # not appear to have issues, but the source is not available yet (as of June 2023).
+  useLLVMStrip = lib.versionAtLeast llvmVersion "15" || lib.versionAtLeast cctoolsVersion "1005.2";
+
+  llvm_bins = [
+    "bitcode_strip"
+    "dwarfdump"
+    "nm"
+    "objdump"
+    "ranlib"
+    "size"
+    "strings"
+  ]
+  ++ lib.optional useLLVMOtool "otool"
+  ++ lib.optional useLLVMStrip "strip";
+
+  # Only include the tools that LLVM doesn’t provide and that are present normally on Darwin.
+  # The only exceptions are the following tools, which should be reevaluated when LLVM is bumped.
+  # - install_name_tool (llvm-objcopy): unrecognized linker commands when building open source CF;
+  # - libtool (llvm-libtool-darwin): not fully compatible when used with xcbuild; and
+  # - lipo (llvm-lipo): crashes when running the LLVM test suite.
+  cctools_bins = [
+    "cmpdylib"
+    "codesign_allocate"
+    "ctf_insert"
+    "install_name_tool"
+    "libtool"
+    "lipo"
+    "nmedit"
+    "pagestuff"
+    "segedit"
+    "vtool"
+  ]
+  ++ lib.optional (!useLLVMOtool) "otool"
+  ++ lib.optional (!useLLVMStrip) "strip";
+
+  ld_path = if useLLD
+    then "${lib.getBin llvmPackages.lld}/bin/ld64.lld"
+    else "${lib.getBin cctools-port}/bin/ld";
+
+  inherit (stdenv.cc) targetPrefix;
+in
+stdenv.mkDerivation {
+  pname = "cctools-llvm-${if useLLD then "lld" else "ld64-${cctoolsVersion}"}";
+  version = llvmVersion;
+
+  nativeBuildInputs = [ makeWrapper ];
+
+  outputs = [ "out" "dev" "man" ];
+
+  buildCommand = ''
+    mkdir -p "$out/bin"
+
+    ln -s "${lib.getDev cctools-port}" "$dev"
+    ln -s "${lib.getMan cctools-port}" "$man"
```

Comment:
Won't this be somewhat misleading? I believe you've mentioned flags that don't correspond for `ranlib`?

Reply:
Libsystem expects to find `include/mach-o` in cctools.dev. I can modify Libsystem to explicitly use `cctools-port.dev` instead to get the headers if that makes more sense. Eventually, it should use cctools-apple, but we’d need to update and make it actually build (which the new version of `ld64` doesn’t even on x86_64-darwin).

For the man pages, I can symlink them from their sources.

Reply:
I made linking the manpages optional, enabled by default but disabled in the stdenv. Linking LLVM’s manpages requires building them, which pulls in a whole bunch of Python stuff.

Reply:
The `dev` part isn`t a problem. It's just taking the cctools man pages even though the binaries are from LLVM that seems like it could potentially cause confusion.

Reply:
The current version of the commit addresses the man pages issue. It takes the cctools ones from cctools and the LLVM ones from LLVM.

I made whether to include them conditional because the LLVM man pages require a bunch of Python dependencies. For the stdenv build, they aren’t built because they’re not included in the output. If a user builds the man pages, they’ll get the mix appropriate to the tools in darwin.cctools-port.out.